source-mongodb/main.go

```diff
@@ -236,44 +298,93 @@ func (d *driver) Validate(ctx context.Context, req *pc.Request_Validate) (*pc.Re
 		return nil, fmt.Errorf("getting list of databases: %w", err)
 	}
 
+	var lastResources []resource
+	var lastBackfillCounters []uint32
+	if req.LastCapture != nil {
+		for _, lastBinding := range req.LastCapture.Bindings {
+			var res resource
+			if err := pf.UnmarshalStrict(lastBinding.ResourceConfigJson, &res); err != nil {
+				return nil, fmt.Errorf("error parsing resource config for lastBinding: %w", err)
+			}
+			lastResources = append(lastResources, res)
+			lastBackfillCounters = append(lastBackfillCounters, lastBinding.Backfill)
+		}
+	}
+
+	var dbCollectionSpecs = make(map[string][]*mongo.CollectionSpecification)
 	var bindings = []*pc.Response_Validated_Binding{}
+	var servInf *serverInfo
 
 	for _, binding := range req.Bindings {
 		var res resource
 		if err := pf.UnmarshalStrict(binding.ResourceConfigJson, &res); err != nil {
 			return nil, fmt.Errorf("error parsing resource config: %w", err)
 		}
 
-		if !slices.Contains(existingDatabases, res.Database) {
-			return nil, fmt.Errorf("database %s does not exist", res.Database)
+		if servInf == nil {
+			servInf, err = getServerInfo(ctx, client, res.Database)
+			if err != nil {
+				return nil, fmt.Errorf("getting server info: %w", err)
+			}
 		}
 
-		var db = client.Database(res.Database)
+		var resourcePath = []string{res.Database, res.Collection}
+
+		// Validate changes to the resource spec.
+		for idx, lastResource := range lastResources {
+			if lastResource.Database == res.Database && lastResource.Collection == res.Collection {
+				// Changing the capture mode is not allowed, although
+				// theoretically going from change stream to a snapshot could
+				// work. It's just easier to not allow it for now unless the
+				// binding is being reset (backfill counter incremented) as
+				// well.
+				if lastResource.getMode() != res.getMode() && binding.Backfill == lastBackfillCounters[idx] {
```

Comment:
Should the backfill counter comparisons here be `binding.Backfill <= lastBackfillCounters[idx]`? Or is the control plane making sure the backfill counter is never decremented & we don't need to check that in the connector?

Reply:
Neither the control plane nor the connector checks that backfill counters are never decremented in general. Materializations do have some code for checking this which I added as a curiosity. It is perhaps debatable if this condition should be enforced. As it is, if you decrement the backfill counter for a task for some reason, the outcome is undefined - it's probably not a good idea, but has actually come in handy sometimes. So I think when designing connectors they should assume that backfill counters always increase, even if that isn't strictly true, and if one happens to decrease we really can't predict what is going to happen and shouldn't try.